core/storage/couchStorageHelpers.go

```diff
@@ -0,0 +1,582 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	_ "github.com/go-kivik/couchdb/v3" // The CouchDB Driver
+	kivik "github.com/go-kivik/kivik/v3"
+	"github.com/open-horizon/edge-sync-service/common"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/open-horizon/edge-utilities/logger"
+	"github.com/open-horizon/edge-utilities/logger/log"
+	"github.com/open-horizon/edge-utilities/logger/trace"
+)
+
+func (store *CouchStorage) checkObjects() {
+	if !store.connected {
+		return
+	}
+
+	var err error
+	currentTime := time.Now().UTC().Format(time.RFC3339)
+	query := map[string]interface{}{
+		"selector": map[string]interface{}{
+			"$and": []map[string]interface{}{
+				map[string]interface{}{"metadata.expiration": map[string]interface{}{"$ne": ""}},
+				map[string]interface{}{"metadata.expiration": map[string]interface{}{"$lte": currentTime}},
+				map[string]interface{}{"$or": []map[string]interface{}{
+					map[string]interface{}{"status": common.NotReadyToSend},
+					map[string]interface{}{"status": common.ReadyToSend}}}},
+		},
+		"fields": []string{"metadata", "last-update"},
+	}
+
+	result := []couchObject{}
+	if err := store.findAll(query, &result); err != nil {
+		if err != notFound && log.IsLogging(logger.ERROR) {
+			log.Error("Error in CouchStorage.checkObjects: failed to remove expired objects. Error: %s\n", err)
+		}
+		return
+	}
+
+	if trace.IsLogging(logger.TRACE) {
+		trace.Trace("Removing expired objects")
+	}
+
+	for _, object := range result {
+
+		id := createObjectCollectionID(object.MetaData.DestOrgID, object.MetaData.ObjectType, object.MetaData.ObjectID)
+		if trace.IsLogging(logger.TRACE) {
+			trace.Trace("Deleting object %s\n", id)
+		}
+
+		var query interface{}
+		query = map[string]interface{}{"selector": map[string]interface{}{"_id": id}}
+		if !object.LastUpdate.IsZero() {
+			query = map[string]interface{}{"selector": map[string]interface{}{"_id": id, "last-update": object.LastUpdate}}
+		}
+
+		if err := store.deleteAllCouchObjects(query); err != nil {
+			if err != notFound || object.LastUpdate.IsZero() {
+				log.Error("Error in CouchStorage.checkObjects: failed to remove expired objects. Error: %s\n", err)
+			}
+		}
+
+		if err == nil {
+			store.DeleteNotificationRecords(object.MetaData.DestOrgID, object.MetaData.ObjectType, object.MetaData.ObjectID, "", "")
+		} else if log.IsLogging(logger.ERROR) {
+			log.Error("Error in CouchStorage.checkObjects: failed to remove expired objects. Error: %s\n", err)
+		}
+	}
+}
+
+func (store *CouchStorage) getOne(id string, result interface{}) common.SyncServiceError {
+
+	db := store.client.DB(context.TODO(), store.loginInfo["dbName"])
+
+	row := db.Get(context.TODO(), id)
+	if kivik.StatusCode(row.Err) == http.StatusNotFound {
```

Comment:
According to the kivik documentation, ['Get()'](https://pkg.go.dev/github.com/go-kivik/kivik/v3#DB.Get) defers all errors until ScanDoc is called. I explicitly checked for the notFound error using row.Err because I was able to replicate and see if this status code was being returned. But I wasn't sure if all errors could be checked this way, so any other error is returned after the ScanDoc call. This should not be a problem according to me...but if this is confusing in terms of code clarity, would you suggest I change it?

Reply:
ok. Yes, this is a surprising behavior, so I would like you to put a comment in the code before every db.Get which just says something simple like "runtime errors are returned on the row.ScanDoc() API. This is the kind of thing that will be forgotten over time.